libsel4vm/src/arch/arm/vgic/vgic.c

```diff
@@ -604,14 +615,17 @@ static int vgic_dist_set_pending_irq(struct vgic_dist_device *d, vm_vcpu_t *vcpu
     virq_data = virq_find_irq_data(vgic, vcpu, irq);
     /* If it is enables, inject the IRQ */
     if (virq_data && gic_dist->enable && is_enabled(gic_dist, irq, vcpu->vcpu_id)) {
-        int err;
         DDIST("Pending set: Inject IRQ from pending set (%d)\n", irq);
 
         set_pending(gic_dist, virq_data->virq, true, vcpu->vcpu_id);
-        err = vgic_vcpu_inject_irq(d, vcpu, virq_data);
-        assert(!err);
-
-        return err;
+        if (vgic_irq_queued(vgic, vcpu, irq)) {
+            return 0;
+        }
+        if (vgic_vcpu_inject_irq(d, vcpu, virq_data)) {
```

Comment:
I consider statements like this very easy to misinterpret, because it's doing the opposite actually.
```suggestion
        int err = vgic_vcpu_inject_irq(d, vcpu, virq_data);
        if (err) { ...
        
```

Reply:
Good idea. Did that here and the other places as well.